# Swift 并发

# 1.概览

## 1. swift并发是什么? 
异步函数+并行代码的组合

## 2. 并发编程的难点在哪里? Swift 并发是如何解决它的?

难点在于:
- 不同运算之间的交互/通信, 按照正确的顺序执行
- 不同运算之间的数据, 如何正确的共享

如何解决?

- 首先Swift5.5 开始支持异步函数 async/await
- 利用结构化并发,保证不同运算的通讯按正确的顺序执行
- 使用actor 保证多线程数据竞争时保证数据的安全

## 3. 那这些到底是什么?

### 3.1 异步函数

```
func loadSignature() async throws -> String? {
let (data, _) = try await URLSession.shared.data(from: someURL) return String(data: data, encoding: .utf8)
}
```

使用规则: await 必须在async函数中调用, async的函数必须要使用await调用

await: 表示函数是可暂停的, 当前线程执行会被暂时用来执行其他任务, 当等待的函数执行完时,会继续向下执行.

### 3.2 结构化并发
对于同步函数来说线程决定了它的执行环境; 对于异步函数来说,我们使用Task, 来完成任务的创建/取消/执行/组织

```
func processFromScratch() async throws { 
    async let loadStrings = loadFromDatabase() 
    async let loadSignature = loadSignature()
    results = []
    let strings = try await loadStrings
    if let signature = try await loadSignature {
      strings.forEach {
        addAppending(signature, to: $0)
    }
} else {
throw NoSignatureError() }
}
```

async let是异步绑定, 被绑定的任务会立刻执行, 即便在await调用之前已经完成, 也要等到await调用才会对其求值.
在这里两个 async let的两个函数会并发执行. 而不是串行.

### 3.3 actor

```
actor Holder {

var results: [String] = []

func setResults(_ results: [String]) {
    self.results = results 
}

func append(_ value: String) { 
    results.append(value)
} 
}
```

用actor标记, 即可简单的实现多线程数据安全.

## 小结

- 异步函数. 异步行为的语法层面支持
- 结构化并发. 提供并发环境, 负责正确的组织/取消/执行 任务
- actor. 封装良好的数据隔离, 确保数据安全

# 2. 创建异步函数

## 使用异步函数的好处
- 避免回调地域
- 合理处理错误, 带有闭包的方法 不会抛出错误
- 避免回调嵌套时, 遗漏分支语句
- 避免隐式的线程调度. 因为在闭包中方法可能会在其他线程调用 进而导致闭包也切换了线程. 但是使用异步函数,当返回暂存点后 还是在当前的线程执行.
- 为swift并发 和 actor 数据安全打下基础

## 异步函数特征
- 和同步函数最大的不同在于, 异步函数会放弃当前线程.
- await 只是提供了一个潜在的暂停点, 而不是必然会暂停, 是否会暂停要看执行环境

## withXXX

如果想要将之前代码中闭包形式的代码 改写成异步函数,可使用下列api



```

func withUnsafeContinuation<T>(
_ fn: (UnsafeContinuation<T, Never>)  Void
) async  T

func withUnsafeContinuation<T>(
_ fn: (UnsafeContinuation<T, Never>) !" Void
) async !" T

func withUnsafeThrowingContinuation<T>(
_ fn: (UnsafeContinuation<T, Error>) !" Void
) async throws !" T

func withCheckedContinuation<T>(
function: String = #function,
_ body: (CheckedContinuation<T, Never>) !" Void
) async !" T


func withCheckedThrowingContinuation<T>( function: String = #function,
_ body: (CheckedContinuation<T, Error>) !" Void
) async throws !" T

```

直观看起来有两种区分维度:
- 是否带throw. 带throw 则可能会抛出异常, 不会则不会
- 是否带unsafe. unsafe则假设开发能正常使用这些api, 如果使用错误 也不会输出异常信息. 

因此: unsafe的性能要高点, 但是如果使用错误 则会出现运行时错误,且没有日志提示.
推荐使用带safe 或者 按debug, release 区分

同时需要注意,使用continuation的resume方法返回结果时,这个方法只能调用一次.
如果需要返回连续的值,则需要使用AsyncStream类的方法

## 异步函数运行环境

异步函数存在潜在的暂停点,因此它的调用必须使用await标记; 而await又必须在async中才能生效.
这就导致异步函数具有很强的传染性.

那最底层(第一个)的异步函数怎么来的?
使用withXXXContinuation包装.

最上层(不想将自己变成async, 又想使用await调用异步函数的函数). 使用Task包装



## 使用异步函数监听通知

```
 Task {
        let backgroundNotifications = await NotificationCenter.default.notifications(
            named: UIApplication.didEnterBackgroundNotification,
            object: nil)
        for await notification in backgroundNotifications {
            
        }
    }
```

这样的写法会使代码更加紧凑, 且对于回调方法也无需添加@objc的表示

但是这样写也会存在显著的问题: 
虽然这样写没有显示的使用self, 但是异步序列还是会持有self,直到序列遍历完. 这样就会导致在序列未完成前, 控制器不会被释放.

解决办法如下:
```
//方法1, 在执行完某次通知后 退出循环
for await notification in backgroundNotifications { print(notification)
    break
}
//但是这样也会有一个问题,就是在期望的通知未执行之前,还是会捕获self

//方法2
let task = Task {
        let backgroundNotifications = await NotificationCenter.default.notifications(
            named: UIApplication.didEnterBackgroundNotification,
            object: nil)
        for await notification in backgroundNotifications {
            
        }
    }
    
//持有task, 在界面退出的时候cancel掉    
task.cancel()

```

# 3. 结构化并发
## 什么是非结构化并发
常见的基于闭包的并发. 它有几个显而易见的缺点:
- 切换了线程,无法获取获取完成的调用栈; 
- 被派发的任务不知道自己的调用者是谁, 无法抛出异常.(只能使用Error/Result代替)
- 导致原本入口和出口明确的函数, 会有多个出口. 进而导致函数结束时, 是否该清理资源变得不那么明确了
- 闭包回调后, 是在哪个环境执行的
- 闭包任务是否可以取消, 假设闭包会二次派发, 那它派发的任务会正常取消吗

## 什么是结构化并发
即使进行并发操作,也要保证控制流入口的单一入口和单一出口. 

即函数中产生的额外控制流路径,最终都会回到函数本身,以保证出口单一.

>为了将并发路径合并，程序需要具有暂停等待其他部分的能力。异步函数恰恰满足了这个条件:
使用异步函数来获取暂停主控制流的能力，函数可以执行其他的异步并发操作并等待它们完成，
最后主控制流和并发控制流统合后，从单一出口返回给调用者。这也是我们在之前就将异步函
数称为结构化并发基础的原因。


## 获取任务属性
```
extension Task where Success !" Never, Failure !" Never { static var isCancelled: Bool { get }
static var currentPriority: TaskPriority { get }
}
```

static var 并不是某个全局属性. 因为一个异步函数 有且仅有一个任务上下文,所有static var获取是当前任务的状态.

结构化并发的任务树和UIView的树形结构很相似,它有以下特性:

- 每个任务都有自己的优先级和取消标识, 每个任务可以有自己的子任务(叶子节点)并在其中执行异步函数
- 当父任务取消时,它的所有子任务也会被取消
- 无论成功还是失败, 子任务的结果都会向上报告给父任务, 在所有子任务完成前/抛出异常前, 父任务是不会被完成.

在任务树上创建子节点,有两种式. async let(异步绑定) 和 async group(任务组)两种方式

## async group

```
struct TaskGroupSample {
    
    func start() async {
        
        print("start")
        
        //Int.self 要求子任务都拥有相同的返回类型,便于后续使用for 循环遍历结果
        //不要在闭包外引用group, 这将会破坏结构化并发的语义,造成未知的后果.
        let result = await withTaskGroup(of: Int.self) { group in
            
            var value = 0
            
            for index in 1...3 {
                group.addTask {
                    //这里是一个并发的环境, 在这里读写value 是不合法的, 会有编译期错误.
                    //为什么是并发的? 你猜它为什么叫结构化并发.
//                    print("value:\(value)")
                    await work(index)
                }
            }
            
            print("task added")
            
            //如果不处理for-loop中的代码,或者只等待部分异步任务就break. 编译器会在检查,在当前group退出前 执行剩下的任务. 自动插入await group.waitForAll()
            for await result in group {
                print("get result: \(result)")
                value = value + result
            }
            
            print("task ended")
            return value
        }
        
        print("end: \(result)")
    }
    
    private func work(_ value: Int) async -> Int {
        print("start work \(value)")
        try! await Task.sleep(nanoseconds: UInt64(value * 2) * NSEC_PER_SEC)
        print("end work \(value)")
        return value
    }
}
```

针对withTaskGroup做以下说明:

- 同一个任务组中的子任务只能拥有同样的返回值类型, 这是为了让TaskGroup可以满足带有强类型的 AsyncSequence 协议所需要的要求(即TaskGroup是一个异步序列). 

- addTask API 把新的任务添加到当前任务中。被添加的任务会在调度器获取到可用资源 后立即开始执行

- group 中的某个任务完成时，它的结果将被放到异步序列的缓冲区中。每当 group 的 next 会被调用时，如果缓冲区里有值，异步序列就将它作为下一个值给出;如 果缓冲区为空，那么就等待下一个任务完成，这是异步序列的标准行为(即异步序列遍历时, 结果的输出顺序由子任务的完成顺序决定, 和子任务的添加顺序无关)。

### TaskGroup 中关键方法签名
创建group
```
@inlinable public func withTaskGroup<ChildTaskResult, GroupResult>(of childTaskResultType: ChildTaskResult.Type, returning returnType: GroupResult.Type = GroupResult.self, body: (inout TaskGroup<ChildTaskResult>) async -> GroupResult) async -> GroupResult where ChildTaskResult : Sendable

```

- ChildTaskResult 是约定了子task的返回值类型
- 整个group的返回值是 GroupResult 类型, 并没有太多约束, 它的值由 body 闭包生成
- group对象是 TaskGroup<ChildTaskResult> 类型, 它满足异步序列协议, 可以使用相关的高阶函数

TaskGroup 添加子任务
``` 
mutating func addTask(
    priority: TaskPriority? = nil,
    operation: @escaping () async -> ChildTaskResult
)
```
TaskGroup的实例方法, 返回值类型由Group创建时的 ChildTaskResult 类型约束.


### TaskGroup的隐式取消
编译器会在检查,在当前group退出前 执行剩下的任务.

### 总结
withTaskGroup 提供了一 种非常 “正规” 的创建结构化并发的方式:它明确地描绘了结构化任务的作用返回，确保在闭包 **内部生成的每个子任务**都在 group 结束时被 await


## async let 异步绑定


```
    func start2() async {
        
        print("start")
        
        //将异步函数绑定到常量vX上, Swift会立即创建一个并发的子任务, async let 赋值后子任务立即执行
        async let v1 = work(1)
        async let v2 = work(2)
        async let v3 = work(3)
        
        print("task added")
        
        //等待执行的结果, await统一等待 和 分别使用三个await 并无差异.
        let result = await v1 + v2 + v3
        print("task end")
        print("result:\(result)")
        
    }
```

async let作用: 创建并发子任务, 绑定到常量上, 并立即执行
await作用: 和其他场景一样,获取异步表达式的值.

### async let的隐式取消
如果作用域结束的时候, 还子任务还没有执行await. 那么编辑器会自动调用取消绑定的任务, 并调用await. 伪代码如下:
```
func start() async { 
    async let v0 = work(0)
    print("End")
    
    // 下面是编译器自动生成的伪代码  注意和 Task group 的不同
    // v0 绑定的任务被取消, 伪代码，实际上绑定中并没有 `task` 这个属性
    v0.task.cancel()
    //隐式 await，满足结构化并发
    _ = await v0
}

```


## 结构化并发的组合
### async let 示例
```
    func start3() async {
        
        async let v02: Int = {
            async let v0 = work(0) //绑定后立即执行
            async let v2 = work(2) //绑定后立即执行
            return await v0 + v2 //等待v0 和 v2的值
        }()
        
        async let v01 = work(1)
        
        _ = await v02 + v01
        
    }
```

需要注意的是, v02 右边是一个匿名异步函数的调用, 它和下面的写法有本质不同
```
async let v02: Int = {
    //这种写法是没有并发的
    return await work(0) + work(2) //串行, 等待v0和v2的值 
}()

```

### TaskGroup 示例
```
    func start4() async {
        await withTaskGroup(of: Int.self) { group in
            
            //等价于上面v02
            group.addTask {
                await withTaskGroup(of: Int.self) { innergroup in
                    innergroup.addTask {
                        await work(0)
                    }
                    
                    innergroup.addTask {
                        await work(1)
                    }
                    
                    return await innergroup.reduce(0) { result, value in
                        return result + value
                    }
                }
            }
            
            //等价于上面v01
            group.addTask {
                await work(1)
            }
            
        }
    }
```
是一点都不直观啊.

### 总结
子任务会从其父任务中继承任务优先级以及任务的本地值 (task local value);在处理任务取消时，除了父任务会将取消传递给子任务外，在子任务中 的抛出也会将取消向上传递

## TaskGroup VS async let
相同点: 
- 创建的任务,都是立即执行的

不同点:
- TaskGroup 要求子任务的返回值类型是一样的, 以方便对异步序列求和. 而async let则灵活的多
- TaskGroup 可以动态生成子任务, 比如for循环动态生成很多子任务; 而async let则是编译期特性, 生成任务个数有限
- 如果在作用域结束的时候,有任务没有处理/调用: TaskGroup 会主动调用await group; 而 async let则是将任务取消 

## 非结构化任务

- TaskGroup.addTask 和 async let - 创建结构化的子任务，继承优先级和本地值。
- Task.init - 创建非结构化的任务根节点，从**当前任务中继承运行环境**:比如 actor 隔离域【举个例子：**控制器中的Task.init继承了main actor的隔离域**，如果想在VC中不继承main actor，则可以使用Task.detached】，优先级和本地值等。
- Task.detached - 创建非结构化的任务根节点，不从当前任务中继承优先级和本地值等运行环境，完全新的游离任务环境。

即通过TaskGroup.addTask 和 async let创建的是结构化的任务, 也仅有的两种创建结构化并发的方式

而Task.init和Task.detached创建的是非结构化的任务.这类任务有以下特性:
- 灵活性最高,可以在任何地方创建,即便是同步函数; 一旦任务被创建,就会被立刻提交并执行
- 每次调用都是生成一个新的任务树(即他们不属于任何任务的子任务),并位于树的顶层.
- 生命周期不和其他作用域绑定,当然也没有结构化并发的特性: 即Task值超过作用域时,不会主动await 也不会自动取消; 想要取消这样的Task的必须手动调用cancel; 外层的cancel 也不会传递到内层(因为他们没有从属关系, 都是顶层任务).

关于是否继承隔离域. 
- 继承(Task.init)即异步函数返回后,还在原来的隔离域.
- 不继承(Task.detached)即不在原来的隔离域

举例1:
```
    func start5() {
        let outer = Task {
            let inner = Task {
                await work(1)
            }
            print("inner.isCancelled:\(inner.isCancelled)") //false innerTask不受外部影响
            await work(2)
        }
        outer.cancel()
        print("outer.isCancelled:\(outer.isCancelled)") //true
    }
```


举例2:
```
    func start6() async {
        
        defer {
            print("start6 结束执行")
        }
        
        async let v1 = Task{
            await work(1)
            print("Cancelled: \(Task.isCancelled)")
        }
        
        async let v2 = Task{
            await work(2)
            print("Cancelled: \(Task.isCancelled)")
        }
        
    }
```

代码输出:
```
//Task创建后立即执行
start work 1
start work 2
//async let 结束执行, 虽然它是结构化的,但是Task.init 创建的任务是非结构化的, 因此它结束的时候 并不能取消非结构化任务
start6 结束执行
//非结构化任务继续执行, 未被取消
end work 1
Cancelled: false
end work 2
Cancelled: false
```

举例3:
```
class MyViewController: UIViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        Task.init {
            //继承隔离域. await返回后还在主线程
            let data = try await URLSession.shared.data(from: URL(string: "")!)
            updateUI(data.0)
        }
        
        Task.detached {
            //继承隔离域. await返回后还不在主线程, 需要使用await
            let data = try await URLSession.shared.data(from: URL(string: "")!)
            await self.updateUI(data.0)
        }
        
    }
    
    @MainActor func explicitUIMethod() {
        
    }
    
    func updateUI(_ data: Any) {
        
    }
    
}

```


### 使用原则
除非有特别的理由，我们希望某个任务独立于结构化并发的生命周期，否则我们应该尽量避免在结构化并发的上下文中使用非结构化任务

# 4. 结构化并发任务的取消
结构化并发的取消,可以很容易得传递到子任务中. 但是:
- 任务取消后, 并不会自动停止.(什么意思, 抛出异常?)
- 任务取消后, 需要手动释放释放的资源 仍然需要处理, 系统不会帮你自动处理

## 那任务取消到底做了什么?
```
    func work() async  -> [Character]{
        var s: [Character] = []
        for c in "Hello" {
            try? await Task.sleep(nanoseconds: UInt64(1) * NSEC_PER_SEC)
            print("current char:\(c), isCancel:\(Task.isCancelled)")
            s.append(c)
        }
        return s
    }
    
    func start7() async {
        let t = Task{
            let result = await work()
            print(String(result))
        }
        
        try? await Task.sleep(nanoseconds: UInt64(2.5 * Double(NSEC_PER_SEC)))
        t.cancel()
    }
```
在2.5s时 停止任务. 但是仍然全部打印了Hello
```
current char:H, isCancel:false
current char:e, isCancel:false
current char:l, isCancel:true
current char:l, isCancel:true
current char:o, isCancel:true
Hello
```

结论:
- cancel 操作将当前task的 isCancelled 标记置为true. 但是任务还是会接着执行
- 在结构化并发中 cancel 当前操作后, 这个动作会传递给它的下级任务
- 外层 group 在接受到子任务的错误后，会主动取消掉任务树中所有的子任务，等待子任务们全部执行完毕 (不论是正常返回还是抛出错误) 后，再进行错误处理

在结构化并发中的取消 被叫做"协作式取消", 即它需要任务树中的所有任务合作,才能完成最终任务的取消.

## 处理任务取消
大致只有两种选择:
- 抛出异常,让当前任务提前结束
- 提前返回一个空值/默认值/部分值

具体使用哪种看情况.

### 抛出异常的方式
```
    func work2() async throws -> [Character]{
        var s: [Character] = []
        for c in "Hello" {
            sleep(1)
            guard !Task.isCancelled else {
                throw CancellationError()
            }
            print("current char:\(c), isCancel:\(Task.isCancelled)")
            s.append(c)
        }
        return s
    }
    
    func start8() async {
        let t = Task{
            do {
                let result = try await work2()
                print(String(result))
            } catch  is CancellationError {
                print("任务被取消")
            } catch {
                print("其他错误")
            }
            
            
        }
        try? await Task.sleep(nanoseconds: UInt64(2.5 * Double(NSEC_PER_SEC)))
        t.cancel()
    }
```

//输入如下内容:
```
current char:H, isCancel:false
current char:e, isCancel:false
任务被取消
```

- CancellationError 是结构性并发中,任务取消对应的错误.
- 检查任务是否取消,并抛出异常十分常见.因此被封装为单独的方法,放到标准库中了. 即使用 try Task.checkCancellation 
 

### 一个稍微复杂点的取消操作
```
    func work3(_ text: String) async throws -> String {
        var s = ""
        
        for c in text {
            
            if Task.isCancelled {
                print("cancelled text:\(text)")
            }
            
            //方法1
//            try Task.checkCancellation() //已经取消 才会抛出异常,否则等下次循环触发
//            sleep(1) //阻塞线程
            
            //方法2
            try await Task.sleep(nanoseconds: NSEC_PER_SEC) //不阻塞线程. 遇到取消时直接抛出异常,不必等到倒计时结束
            print("Append char:\(c), isCancelled:\(Task.isCancelled)")
            
            s.append(c)
        }
        
        return s
    }
    
    func start9() async {
        
        do {
            let value: String =  try await withThrowingTaskGroup(of: String.self) { group in
                
                //task1
                group.addTask {
                    
                    try await withThrowingTaskGroup(of: String.self) { innerGroup in
                        //task 1.1
                        innerGroup.addTask{
                            try await work3("hello")
                        }
                        
                        //task 1.2
                        innerGroup.addTask {
                            try await work3("world")
                        }
                        
                        try await Task.sleep(nanoseconds: UInt64(2.5 * Double(NSEC_PER_SEC)))
                        innerGroup.cancelAll()
                        return try await innerGroup.reduce([]) { $0 + [$1] }.joined(separator: " ")
                    }
                }
                
                //task2
                group.addTask {
                    try await work3("swift concurrency")
                }
                
                return try await group.reduce([]) { $0 + [$1] }.joined(separator: " ")
            }
            
            print("value:\(value)")
            
        } catch {
            //try-catch 中error会向上传递, 可以只在最外层捕捉, 简化代码处理.
            print("error:\(error)")
        }
        
    }
```

## 取消的清理工作

### 使用defer
当前代码快的作用域结束时 或者 抛出了异常,就会触发defer
注意defer要写对地方

### Cancellation Handler
使用Task.checkCancellation 抛出取消异常, 要比取消操作要完一些, 因为它常常等到下一次事件才会触发

更好的方法是使用Cancellation Handler, 好处
- 及时,异常抛出立马处理
- 可以处理同步函数使用续体包装的异常.

```
func withTaskCancellationHandler<T>(
    handler: () -> Void,
    operation: () async throws -> T
) async rethrows -> T
```

# 5. actor 模型和数据隔离

## 现有多线程和锁的设计问题

多线程问题:
每个线程有自己的栈空间, 但是不同线程公用堆上的空间. 因此会导致多个线程同时操作一份数据的情况.

锁的问题:
锁可以解决多线程同时操作数据的问题,但是它自身也有一定的问题:

- 没有拿到锁的线程,只能无意义的等待,做不了任何事情. 这限制了多线程并发的性能
- 在锁的使用,如果全局只用一把(少量的)锁, 这会造成很多等待; 而锁一旦多起来, 又有造成死锁的风险

## actor 隔离

1. actor是什么?
以实例为单位定义了一个隔离域, 域内的以同步方式访问, 域外以异步的方式访问. 无论域内外, 是否可重入, 对actor的属性/方法的操作 都是串行的.


2. actor的作用
- 每一个actor上的状态修改都是顺序执行的
- 对属性的直接访问/操作, 只能存在于actor内部.
- actor隔离: 对于actor实例上的所有内容,如计算属性/存储属性,实例方法, 都是actor隔离的. 即便这些实例上的操作没有被标记为async,也只能通过await 调用. 这就要求了隔离于上的调用,必须在异步环境的上下文中

3. actor隔离域?
可以理解为actor实例中的变量/行为,处于一个串行队列中? 或者说对于actor实例中所有的变量和行为,都必须使用await调用.

4. nonisolated标记的成员,无法访问隔离域内的成员.

5.一些结论
- 在 actor 中的声明，默认处于 actor 的隔离域中。
- 在 actor 中的声明，如果加上 nonisolated，那么它将被置于隔离域外。
- 在 actor 外的声明，默认处于 actor 隔离域外。

针对第三点, 当actor描述的对象 当做参数传递时, 可以使用 Isolated 关键字, 使其处于隔离域内. 如
```

func reportRoom(room: isolated Room) { 
    print("Room: \(room.visitorCount)")
}
```

但是这样呢? 
```
//在Swift6 这是被编译器禁止的 
    func addCount(room1: isolated Room, room2: isolated Room) -> Int {
        let count = room1.visitorCount + room2.visitorCount
        return count
    }
    
    //在swift5.x 会直接报错
    func baz(_ another: isolated Room) { 
        print(self.visitorCount)
        print(another.visitorCount)
    }
```


举个例子:
```
actor Room {
    let roomNumber = "101" 
    
    var visitorCount: Int = 0
    
    init() {}
    
    func visit() -> Int {
        visitorCount += 1
        return visitorCount
    }
}

protocol Popular {
    var popular: Bool { get }
}


extension Room: Popular {
    var popular: Bool {
        return visitorCount > 10
    }
}
```

Room是actor隔离的, 所有的属性/方法调用都必须使用await, 包括Room.popular
但是Popular.popular 没有这个要求,它只是一个同步的计算属性.
这就造成了语义上的冲突.

因此编译器会报错:
```
Actor-isolated property 'popular' cannot be used to satisfy nonisolated protocol requirement
```

解决办法:
```
//方法1. 将Popular变成actor隔离的
protocol PopularActor: Actor {
    var popularActor: Bool { get }
}

//方法2  将popularAsync属性变成 异步的, 让它在语法上满足可暂停的特点
protocol PopularAsync {
    var popularAsync: Bool { get async }
}

extension Room: Popular {
    var popular: Bool {
        //实现1. 这里也不会报错, 那这个实现是否是线程安全的?
       // return visitorCount > 10
        //实现2 使用了async 虽然满足了协议,但是使用了async 有很强的传染性
    get async {
        return visitorCount > 10
        }
    }
}

```

# 6. 全局actor/可重入/Sendable

## 全局actor

### 何时需要全局的actor?
- 代码不能汇集到单一actor时, 就需要作用域更大的actor
- 代码比较分散时

作用: 所有被属性包装标注的内容，都会被限制在这个全局 actor 单例的隔离域中

### 如何定义
```
@globalActor actor MyActor { 
    static let shared = MyActor()
    private init() { }
}
```

通常不在 @globalActor 里持有变量和实例方法，并且为它声明一个私有的初始化方法。这样，我们就能将这个 actor 类型作为纯粹的标记来使用，减少一些迷惑.

### 如何使用全局actor?
任何需要使用它的地方加上 注解即可

@MyActor var foo = "some value"

### MainActor

作用: 判断当前线程是否在主线程,如果在直接执行代码;如果不在,切换到主线程再执行.

任意需要主线程的地方,使用MainActor标记就行了. 其他全局actor使用也一样.

```
    @MainActor class C1 { 
        func method() {
            
        }
    }
    class C2 {
        @MainActor var value: Int?
        @MainActor func method() {
            
        }
        func nonisolatedMethod() {
            
        }
    }
    @MainActor var globalValue: String = ""
```

编译检查:
 @MainActor 进行隔离域检查只发生在异步方法里、actor中、以及 Task 相关 API 等地方.
 同步方法不检查, 如果检查代码大概率编译不过
 


## 可重入

### 出现场景
actor 中使用异步函数

### 定义
可重入:在被暂停的 actor 隔离函数继续 之前，可重入特性允许其他工作在这个 actor 上执行, 甚至是再次调用已暂停的函数

所以将actor理解为串行队列也是不准确的，因此在actor中的异步函数暂停时，其他工作仍然可以在这个actor上执行，甚至是再次执行这个暂停的函数.

但是要注意,虽然可重入, 但是在隔离域内依赖是串行的.

### 有什么问题
被暂停的函数,await前后执行环境 可能发生变化.
 
举例:
```
    func start10() {
        let room = Room()
        //task1
        Task {
            await room.visit()
            let r = await room.generateReport() //延迟1s执行
            print("analyze result:\(String(describing: r))")
        }
        
        //task2
        Task {
            for _ in 0...100 {
                await room.visit() //无延迟. 加await只是因为不在隔离域.
            }
        }
        
        //在task 1等待期间, task2 仍然可以执行.
    }
    
    
    actor Room {
    let roomNumber = "101"
    
    var visitorCount: Int = 0
    
    init() {}
    
    func visit() -> Int {
        visitorCount += 1
        return visitorCount
    }
    
    //actor内的异步函数
    func generateReport() async -> Result? {
        
        if !popular { //这里依赖visitorCount, 在异步函数返回后它的值已经变了
            let reason = await analyze(room: self)
            return Result(reason: reason, count: visitorCount)
        } else {
            return nil
        }
        
    }
    
        var popular: Bool {
        
                visitorCount > 10
    }
}

```

### 既然会有问题, 那为什么还要有?
如果不允许重入, 就和锁一样了, 不能发挥最大限度发挥并发的威力

### 如何解决重入问题?
如果是值类型, 在await前 copy一份




## sendable

### 什么是sendable

在不同并发域安全传递数据的能力. 仅编译器有效, 运行时什么也不做

### 为什么需要sendable

因为actor虽然能保证在自己的成员是线程安全的,但不能保证跨域参数/返回值是线程安全的


### 什么样的数据,可以在不同的隔离域传递?
- 值类型, 且所有成员都满足值语义
- 引用类型, 成员是let & 满足sendable
- 内部有保护机制, 比如actor 或者 锁
- 深拷贝传递的

### class 类型, 如何满足sendable
- class 是final的
- class的所有成员是sendable的
- 所有成员是let的

### actor 类型
actor 也是引用了类型; actor 可以随意在不同并发域传递

### 函数和@sendable
函数也是引用类型, 在跨越并发域时，在函数体内的这些被引用的值可能会发生变化。想要保 证数据安全的话，我们必须规定函数闭包所持有的值都满足 Sendable 且它们都为不变量.

### @unchecked Sendable
如果确认时线程安全的, 可以使用这个标记 绕过编译检查 

# 7. 并发线程模型

## 协同式线程池

### 原理
swift并发 在底层使用的是一种协同式线程池的调用方式:
- 有一个串行队列,负责将异步函数转换为续体来进行调度
- 协同式式线程池, 系统创建的不多于CPU核心数的线程池
- 一个异步函数的续体,可能会执行在不同的线程上.

所以要避免阻塞调度队列,如果阻塞了,那就无法进行任务派发了,更执行不了了.

一般流程:
1.任务被分割为续体，同一个任务的不同续体，甚至是同一个续体可能是在不同线程执行的
2.执行信息/过程被记录到堆上，当有可用线程时 会把信息恢复到栈上
3.队列不会等待，它一直处于寻找任务处理的状态



### 传统线程模型的困境

会创建很多个线程:
因为当线程被阻塞或者未拿到锁时, 线程处于不可分配的状态, 当有新的任务来临时, 系统倾向于创建新的线程

而当线程过多时, 首先它需要更多的内存,其次线程切换也是耗时的.

这也是线程爆炸的根本原因: 串行队列的线程被阻塞.

### 为什么要使用协同式线程池?
它要提供一种和线程切换不同的方式,要更轻量级. 续体间切换,和普通的方法调用可以等价. 减少线程切换的耗时

一般协同式线程的数量不多于CPU核心数， 这样就

### 如何正确使用它
避免调度队列被阻塞:
因为一旦阻塞就会创建更多的线程. 

- 运行在这个线程上的异步函数,要有放弃线程的能力.[这个就是关键,要有放弃线程的能力.]
- 避免使用阻塞式API
- await的一边 成对的调用锁

### 错误示范
```
func blockingMethod() async -> Bool {
    Thread.sleep(forTimeInterval: 1.0)
    return true
}
```

虽然是异步函数,但是使用了阻塞式API, 会影响队列调度.



# 8.回顾
Swift 并发主要是解决什么问题？

- 以async/await为代表的异步函数，具有放弃当前的能力，且异步函数在底层表示一个个可独立执行的续体。相对于线程，续体更为轻量 在不同续体间切换时性能更好。同时异步函数，避免了回调地域，统一了函数出口
- task相关的概念 提供了并发能力
- actor 相关概念，保证了并发时数据的安全。
- 编译时 暴露问题

有了线程和进行，为什么还需要协程？
线程如果使用不当，会造成线程爆炸，占用大量的内存，且线程切换时也是耗时的。

协程的最大优势就是：切换比线程快。 因为它轻量啊，执行的是一个个续体

协程最大的特点：它可以暂停和挂起。

协程的线程池，一般不多于CPU核心数， 这几个线程在不同的续体且切换来执行任务。因此它避免线程爆炸和线程切换耗时。 

但同时也要避免在异步函数中执行阻塞线程的任务，因为这会阻塞协程。慎用锁、信号量


纸上得来终觉浅，后面不要在写闭包回调了，多用用体会下。

# 9.一些测试

## 1. 运行线程
```
func test0() {
        DispatchQueue.global().async {
            DispatchQueue.main.async {
                self.test1()
            }
        }
    }
    
    func test1() {
        print("check1: \(Thread.isMainThread)")
        //Thread 2 Queue : com.apple.root.user-initiated-qos.cooperative (concurrent)

        Task {
            print("check4: \(Thread.current)")
            let result = await fetchDataAsync()
            print("check2: \(Thread.current), res:\(result)")
        }
            
        print("check3: \(Thread.isMainThread)")
    }
```

test1如果运行在UIViewController中 那么check2和4 都在主线程(因为继承了隔离域)。
但如果在其他自定义类中，那么事情将变得有趣
```
check1: true
check3: true
check4: <NSThread: 0x60000170afc0>{number = 6, name = (null)}
check2: <NSThread: 0x60000170aec0>{number = 5, name = (null)}, res:Data received
```
可以看到await前后是在不同的线程。且Task中的任务 并不是在test1所在的主线程执行。

- Task中的任务在系统维护的协同式线程池执行，而不是原来的运行环境
- Task中的多个异步函数、或者同一个异步函数的多个续体，并不保证是在同一个线程执行
- Task中的任务是顺序执行的

为什么这样做？
避免线程爆炸，以及线程切换时耗时
